Add CLion to "Tools and IDEs" (#60204)
zephyrproject-rtos/zephyr: doc/develop/tools/clion.rst

.. _clion_ide:

..
   Copyright 2023 Marina Kalashina <marina.kalashina@jetbrains.com>

CLion
#####

`CLion <https://www.jetbrains.com/clion/>`_ is a cross-platform C/C++ IDE that
supports multi-threaded RTOS debugging.

This guide describes the process of setting up, building, and debugging
Zephyr's `threads` sample in CLion.

Get CLion
*********

Download_ and install CLion.

.. _Download: https://www.jetbrains.com/clion/download

Initialize a new workspace
**************************

This guide gives details on how to build and debug the
`threads` sample application, but the instructions would
be similar for any Zephyr project and
:ref:`workspace layout <west-workspaces>`.

Before you start, make sure you have a working Zephyr development
environment, as per the instructions in the :ref:`getting_started`.

Open the project in CLion
**************************

> #. In CLion, click **Open** on the Welcome screen or select **File | Open**

Reviewer: Please look at implementing similar changes for all the other GUI and menu elements mentioned in the page, to keep things more consistent.

   from the main menu.

#. Locate your Zephyr workspace folder (``~/zephyrproject`` if you
   followed the Getting Started instructions), then select ``~/zephyrproject/zephyr/samples/basic/threads``
   or another sample project folder.

   Click :guilabel:`OK`.

#. If prompted, click :guilabel:`Trust Project`.

   See `this section`_ in CLion web help for more information on project
   security.

   .. _this section: https://www.jetbrains.com/help/clion/project-security.html#projects_security

Configure the toolchain and CMake profile
*****************************************

> #. In the **Open Project Wizard**, click :guilabel:`Add profile`.

Author: Thank you for spotting this, indeed the wizard opens right at the profile setting and there's no need to add a new profile. Corrected.

#. If the toolchain is not yet configured, click :guilabel:`Manage Toolchains` next

>    to the **Toolchain** field. We recommend that you use MinGW with default

Reviewer: or maybe this should actually be as follows, as per the screen capture and if this entry is available by default for all Windows users?

```suggestion
   to the :guilabel:`Toolchain` field. We recommend that you use :guilabel:`Bundled MinGW` with default
```

>    to the **Toolchain** field. We recommend that you use MinGW with default

Author: That's right, Bundled MinGW is available for all Windows users. I also highlighted 'System' since this is how this default macOS/Linux toolchain is called in CLion.

   settings on Windows or the System toolchain on Unix machines.

> #. In the toolchain settings, click :guilabel:`Add environment` and select ``~/zephyrproject/.venv/Scripts/activate.bat``.

Reviewer: also:

```suggestion
#. In the toolchain settings dialog (:guilabel:`Manage toolchaines...`), click :menuselection:`Add environment --> From file` and select ``~/zephyrproject/.venv/Scripts/activate.bat``.
```

> #. In the toolchain settings, click :guilabel:`Add environment` and select ``~/zephyrproject/.venv/Scripts/activate.bat``.

Author: We've managed to check 'activate.bat' on Linux and it works, but for the case of macOS, we need to wait for our engineer who is familiar with Zephyr setup on macOS to take a look and help with that. This will next year :)
I added a disclaimer about the OSs in the beginning, so maybe we could leave this section as is for now, and add a note on macOS specifics later? Would that be ok?

> #. In the toolchain settings, click :guilabel:`Add environment` and select ``~/zephyrproject/.venv/Scripts/activate.bat``.

Reviewer: Yes this sounds good, thanks!

>    .. image:: img/toolchain_mingw.webp
>       :width: 600px
>       :alt: MinGW toolchain with environment script

Reviewer: please update images to figures so that there is a caption. Also, use center alignment.

```suggestion
   .. figure:: img/toolchain_mingw.webp
      :width: 600px
      :align: center
      :alt: MinGW toolchain with environment script

      MinGW toolchain with environment script
```

   Save the changes.

> #. Back in the CMake profile settings, specify your board in the :guilabel:`CMake options` field.

Reviewer: ```suggestion
#. Back in the CMake profile settings dialog, specify your board in the :guilabel:`CMake options` field.
```

   See the `Project security`_ section in CLion web help for more information on project security.

Configure the toolchain and CMake profile
*****************************************

CLion will open the :guilabel:`Open Project Wizard` with the CMake profile settings. If that does
not happen, go to :menuselection:`Settings --> Build, Execution, Deployment --> CMake`.

#. Click :guilabel:`Manage Toolchains` next to the :guilabel:`Toolchain` field. This will open the
   :guilabel:`Toolchain` settings dialog.

#. We recommend that you use the :guilabel:`Bundled MinGW` toolchain with default settings on
   Windows, or the :guilabel:`System` (default) toolchain on Unix machines.

#. Click :menuselection:`Add environment --> From file` and select
   ``..\.venv\Scripts\activate.bat``.

   .. figure:: img/toolchain_mingw.webp
      :width: 600px
      :align: center
      :alt: MinGW toolchain with environment script

   Click :guilabel:`Apply` to save the changes.

#. Back in the CMake profile settings dialog, specify your board in the :guilabel:`CMake options`
   field. For example:

   .. code-block::

      -DBOARD=nrf52840dk/nrf52840

>    .. figure:: img/cmakeprofile.webp

Reviewer: screen capture needs update w.r.t. new board name (nrf52840dk/nrf52840)

> .. _CLion:

Reviewer: ```suggestion
.. _clion:
```

Otherwise this clashes with the header defined at line 7.

..
   Copyright 2023 Marina Kalashina <marina.kalashina@jetbrains.com>

CLion
#####

`CLion <https://www.jetbrains.com/clion/>`_ is a cross-platform C/C++ IDE that
supports multi-threaded RTOS debugging.

> This guide describes the process of setting up, building, and debugging
> Zephyr's example-application_ in CLion.

Reviewer: ```suggestion
This guide describes the process of setting up, building, and debugging
Zephyr's :ref:`hello_world` sample in CLion.
```

> .. _example-application: https://github.com/zephyrproject-rtos/example-application
> 
> 

Reviewer: ```suggestion
```

Initialize a new workspace
**************************

#. Before you start, make sure to prepare a Zephyr development environment:

   * `Install dependencies`_

      .. _Install dependencies: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#install-dependencies

   * `Get Zephyr and install Python dependencies`_

      .. _Get Zephyr and install Python dependencies: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#get-zephyr-and-install-python-dependencies

   * `Install Zephyr SDK`_

      .. _Install Zephyr SDK: https://docs.zephyrproject.org/latest/develop/getting_started/index.html#install-zephyr-sdk

#. Run the following command to initialize the workspace folder
   (``my-workspace``) and clone the example application into it:

   .. code-block::

      west init -m https://github.com/zephyrproject-rtos/example-application
      --mr main my-workspace

#. Update Zephyr modules:

   .. code-block::

      cd my-workspace
      west update

Open the project in CLion
**************************

#. In CLion, click **Open** on the Welcome screen or select **File | Open**
   from the main menu.

> #. Locate the workspace folder (``my-workspace``) and select it.

Author: I recall we tried opening a project instead of workspace, but that didn't work for us, unfortunately. So since we couldn't test that properly, we decided not to include that in the instructions and only write about what actually worked.

   Click **OK**.

> #. If prompted, click **Trust Project**.

Author: Added a recommendation to use MinGW or 'System' (on unix)

   See `this section`_ in CLion web help for more information on project
   security.

   .. _this section: https://www.jetbrains.com/help/clion/project-security.html#projects_security
    </chapter>

Create a CMake presets file
***************************

This presets_ file will activate the Python virtual environment for your Zephyr
installation and will define the target board, the build directory,

> and the flashing tool.

Author: The preset activates venv for CMake builds, but indeed not for the terminal window. Changed into "This presets file will activate the Python virtual environment and will define the target board for CMake builds."

.. _presets: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html

You can find more information on how CLion works with CMake presets on
`this web help page`_.

.. _this web help page: https://www.jetbrains.com/help/clion/cmake-presets.html

> #. In the **Project** view, locate the ``my-workspace\example-application\app``

Reviewer: ```suggestion
#. In the **Project** view, locate the ``zephyrproject\zephyr\samples\hello_world``
```

   folder.

> #. Right-click the ``app`` folder and select **New | File**.

Reviewer: ```suggestion
#. Right-click the ``hello_world`` folder and select **New | File**.
```

#. Name the file ``CMakeUserPresets.json``.

#. Add the following code into it:

   .. code-block::

      {
         "version": 3,
            "cmakeMinimumRequired": {
            "major": 3,
            "minor": 24
         },
         "configurePresets": [
            {
               "name": "Zephyr-Debug",
               "generator": "Ninja",
               "binaryDir": "build",
               "cacheVariables": {
                  "BOARD": "nucleo_f302r8",

>                   "CMAKE_PROGRAM_PATH": "&lt;OPENOCD_INSTALLATION_PATH&gt;;&lt;HOME_PATH&gt;/zephyrproject/.venv/Scripts"

Reviewer: Don't use &lt; / &gt; as it's already a code block, plus, please explain better in the paragraph that follows that <SOME_TAGS> need to be replaced by actual values, as otherwise the fact they are written in capital letters makes it look like it might automagically work for these could be existing environment variables.

Author: Removed &lt; / &gt; and capitalization. Below we instruct to customize that line, is it clear what should be changed there?

               }
            }
         ],
         "buildPresets": [
            {
               "name": "Zephyr-Debug",
               "configurePreset": "Zephyr-Debug"
            }
         ]
      }

#. Customize the following lines:

   * ``"BOARD":...`` - specify the board you are using.

   * ``"CMAKE_PROGRAM_PATH":...`` - set the path to Zephyr environment scripts
     (delete this part if you are not using a virtual environment) and set
     the path to your flashing software (OpenOCD in our example).

#. Save the file. Make sure it is located in the ``app`` folder.

Load CMake
**********

#. In the **Project** tree, locate the ``CMakeLists.txt`` file in the ``app``
   folder.

#. Right-click it and select **Load CMake Project**.

>    CMake load should finish successfully.

Author: Which dialog, is it about the profiles?

>    CMake load should finish successfully.

Reviewer: can't remember, resolving for now :)

Configure CMake profiles
************************

#. Go to **Settings | Build, Execution, Deployment | CMake**.

> #. Delete the automatically created **Debug** profile.

Author: Added an intro at the top, also switched the steps 2 and 3 and gave an alternative about deleting the Debug profile or moving the preset profile to the top instead. Does it look more clear now?

#. Select the **Zephyr-Debug preset** profile and set the **Enable profile**
   checkbox.

#. Click **Apply** and close the dialog.

   CMake project will be reloaded.

Build the project
*****************

#. Check the top toolbar: the **Zephyr-Debug** profile and the
   **zephyr_final** configuration should be selected.

#. Click the hammer icon next to the configuration switcher.

   When the build is finished, you can find build artifacts in
   the ``app\build`` folder.

> 

Reviewer: Would be worth highlighting that other CMake targets can now of course be called, beyond just build (ex. `puncover`, `hardenconfig`, etc.)

Author: Added a sentence on that

   This option enables Zephyr tasks view during debugging. See
   `Multi-threaded RTOS debug`_ in CLion web help for more information.

   .. _Multi-threaded RTOS debug: https://www.jetbrains.com/help/clion/rtos-debug.html

Create an Embedded GDB Server configuration
*******************************************

#. In the configuration switcher, select **Edit Configurations**.

#. Click "+" in the left-hand pane and select **Embedded GDB Server**.

#. Configure the settings:

   * **Name** (optional): Debug Zephyr

   * **Target**: zephyr_final

   * **Executable binary**: zephyr_final

   * **Debugger**: Bundled GDB

      *Note*: For non-ARM and non-x86 architectures, use a GDB executable form
      Zephyr SDK. Make sure to pick a version with Python support (for
      example, **riscv64-zephyr-elf-gdb-py**) and check that Python is
      presented in the system ``PATH``.

   * **Download executable** (optional): specify when to download the
      executable binary to the target device.

   * **'target remote' args** (optional): provide the GDB server-specific
      arguments (for example, port number or board config file).

   * **GDB Server**: cmake

   * **GDB Server args**:

>       ``--build --preset Zephyr-Debug --target flash debugserver``

Author: Quoting @elmot (for some reason, his last comments got 'pending' or don't appear in the thread at all): the most reliable way is to keep 'flash' here and set 'Download executable' to Never

>    * **Working directory**: the path to ``app``

Reviewer: ```suggestion
   * **Working directory**: Directory where `CMakeUserSettings.json` is located so that CMake can find the Zephyr-Debug CMake preset, i.e. ``app``.
```

Also, could this be set using a macro (clicking the "+") instead? having to navigate to the folder and set it as an absolute path seems wrong? `$CMakeCurrentBuildDir$/..` works, but not sure if there is a better one?

Author: Unfortunately, some macros can be buggy, so it's safer to ask for the full path

supports multi-threaded RTOS debugging.

This guide describes the process of setting up, building, and debugging
Zephyr's :ref:`hello_world` sample in CLion.

Get CLion
*********

Download_ and install CLion.

.. _Download: https://www.jetbrains.com/clion/download

Initialize a new workspace
**************************

This guide gives details on how to build and debug the
:ref:`hello_world` sample application, but the instructions would
be similar for any Zephyr project and
:ref:`workspace layout west-workspaces`.

Before you start, make sure you have a working Zephyr development
environment, as per the instructions in the :ref:`getting_started`.

Open the project in CLion
**************************

#. In CLion, click **Open** on the Welcome screen or select **File | Open**
   from the main menu.

#. Locate your Zephyr workspace folder (``~/zephyrproject`` if you
    followed the Getting Started instructions) and select it.

   Click **OK**.

#. If prompted, click **Trust Project**.

   See `this section`_ in CLion web help for more information on project
   security.

   .. _this section: https://www.jetbrains.com/help/clion/project-security.html#projects_security

>     </chapter>

Reviewer: remove??

```suggestion
    </chapter>
```

> #. If the toolchain is not yet configured, CLion will open the

Reviewer: wouldn't we want to have folks use the Zephyr toolchain instead? or maybe it doesn't matter as west/ninja then take over anyway?

Reviewer 2: Might be it's better to add that the compilers do not matter, zephyr uses it's own SDK anyway

    Open Project Wizard. We recommend that you use MinGW with default
    settings on Windows or the System toolchain on Unix machines.

Create a CMake presets file
***************************

This presets_ file will activate the Python virtual environment
and will define the target board for CMake builds.

.. _presets: https://cmake.org/cmake/help/latest/manual/cmake-presets.7.html

You can find more information on how CLion works with CMake presets on
`this web help page`_.

.. _this web help page: https://www.jetbrains.com/help/clion/cmake-presets.html

#. In the **Project** view, locate the
    ``zephyrproject\zephyr\samples\hello_world`` folder.

#. Right-click the ``hello_world`` folder and select **New | File**.

#. Name the file ``CMakeUserPresets.json``.

#. Add the following code into it:

   .. code-block::

>       :emphasize-lines: 13,14

Reviewer 3: Is this the reason the code block is not being shown in the preview? Because I do not see it.

      {
         "version": 3,
            "cmakeMinimumRequired": {
            "major": 3,
            "minor": 24
         },
         "configurePresets": [
            {
               "name": "Zephyr-Debug",
               "generator": "Ninja",
               "binaryDir": "build",
               "cacheVariables": {
                  "BOARD": "nucleo_f302r8",
                  "CMAKE_PROGRAM_PATH": "<openocd_installation>;<home>/zephyrproject/.venv/Scripts"
               }
            }
         ],
         "buildPresets": [
            {
               "name": "Zephyr-Debug",
               "configurePreset": "Zephyr-Debug"
            }
         ]
      }

> 

Reviewer 3: I have tried this Configure Preset, but for me it generates 2 presets: 1 Zephyr-Debug and a Zephyr-Debug-Zephyr-Debug
First one does not have "Build options" set second one has a "Build options": "--preset "Zephyr-Debug""

If I want some overlays and or config settings in these presets, how should I add these?

Reviewer 3: Would it also be an option to add a preset for native_posix for example. This to see how you should work with multiple platforms. Would be a big plus for me. I can help with testing and trying out if you want.
I personally would love to get it working for nrf53 and posix. Always have been struggling with CLion.

Reviewer 2: Do I understand correctly that `native_posix` and `native_posix_64` are not supported on windows?

Reviewer 2: @MarinaKalashina I'd also add `.gitignore`:
```gitignore
/cmake-build*/
/build*/
CMakeUserPresets.json
```

#. Customize the following lines:

   * ``"BOARD":...`` - specify the board you are using.

   * ``"CMAKE_PROGRAM_PATH":...`` - set the path to Zephyr environment scripts
     (delete this part if you are not using a virtual environment) and set
     the path to your flashing software (OpenOCD in our example).

#. Save the file. Make sure it is located in the ``hello_world`` folder,
    i.e. the same folder as your CMake project.

Load CMake
**********

#. In the **Project** tree, locate the ``CMakeLists.txt`` file in the

>     ``hello_world`` folder.

Reviewer: ```suggestion
   ``hello_world`` folder.
```

BTW, you can use 100-character long lines if you want, no need to wrap at 80.

#. Right-click it and select **Load CMake Project**.

   CMake load should finish successfully.

Configure CMake profiles
************************

CLion automatically creates a CMake profile for any detected preset,
but such profiles are disabled by default. Follow the steps below to
enable the profile associated with our preset and set it as a working
profile for the project.

#. Go to **Settings | Build, Execution, Deployment | CMake**.

#. Select the **Zephyr-Debug preset** profile and set the **Enable profile**
   checkbox.

#. Delete the automatically created **Debug** profile. Alternatively,
    move the **Zephyr-Debug preset** profile to the top of the list
    to make it default for the project.

#. Click **Apply** and close the dialog.

   CMake project will be reloaded.

Build the project
*****************

#. Check the top toolbar: the **Zephyr-Debug** profile and the

>    **zephyr_final** configuration should be selected.

Reviewer 3: actually I don't have zephyr_final, I only have zephyr and zephyr_pre0 (and bunch of other of course, but no zephyr_final)
but I altered these lines: 
```
        "BOARD": "native_posix_64",
        "CMAKE_PROGRAM_PATH": "<home>/zephyrproject/.venv/Scripts"
```
since I don't use openOCD for posix or nRF53

#. Click the hammer icon next to the configuration switcher.

   When the build is finished, you can find build artifacts in
   the ``hello_world\build`` folder.

Note that other CMake targets like ``puncover`` or ``hardenconfig``
can also be called at this point.

Configure Zephyr parameters for debug
*************************************

#. In the configuration switcher, select **guiconfig** and click the hammer
   icon.

#. Use the GUI application to set the following flags:

   .. code-block::

>       CONFIG_DEBUG_THREAD_INFO

Reviewer 3: Atcually you need to remove the CONFIG_ to search for it.

Author: @kartben @lucsegers Thank you! Sorry I missed that in the first pass

Reviewer: No worries! I have started another (I think final!) round of reviews and I'm pretty confident we can get this in by the end of the year 🙃 thanks again for all the efforts on this one

      CONFIG_THREAD_RUNTIME_STATS
      CONFIG_DEBUG_OPTIMIZATIONS

Enable RTOS integration
***********************

#. Go to **Settings | Build, Execution, Deployment | Embedded Development |
   RTOS Integration**.

#. Set the **Enable RTOS Integration** checkbox.

   This option enables Zephyr tasks view during debugging. See
   `Multi-threaded RTOS debug`_ in CLion web help for more information.

   .. _Multi-threaded RTOS debug: https://www.jetbrains.com/help/clion/rtos-debug.html

    You can leave the option set to **Auto**. CLion will detect Zephyr
    automatically.

Create an Embedded GDB Server configuration
*******************************************

In order to debug the Zephyr application, you need to create
a run/debug configuration out of the Embedded GDB Server template.

> This configuration will call :ref:`west debugserver` under the hood

Reviewer: you may add a blank line to actually make it two paragraphs

and then use CLion's embedded GDB server to flash the target and
start a debug session.

#. In the configuration switcher, select **Edit Configurations**.

#. Click "+" in the left-hand pane and select **Embedded GDB Server**.

#. Configure the settings:

    .. list-table::
        :header-rows: 1

        * - Option
          - Value

        * - **Name** (optional)
          - Debug Zephyr

        * - **Target**
          - zephyr_final

        * - **Executable binary**
          - zephyr_final

        * - **Debugger**
          - Bundled GDB

>             *Note*: For non-ARM and non-x86 architectures, use a GDB executable
>             form Zephyr SDK. Make sure to pick a version with Python support
>             (for example, **riscv64-zephyr-elf-gdb-py**) and check that Python
>             is presented in the system ``PATH``.

Reviewer: maybe use note admonition - it doesn't render 100% perfectly, but it might be worth using it nevertheless if you really want to make sure people don't miss out this piece of information.

```suggestion
            .. note:: For non-ARM and non-x86 architectures, use a GDB executable
               form Zephyr SDK. Make sure to pick a version with Python support
               (for example, **riscv64-zephyr-elf-gdb-py**) and check that Python
               is presented in the system ``PATH``.
```
